types/src/v0/impls/state.rs

```diff
@@ -283,32 +303,41 @@ impl From<MerkleTreeError> for FeeError {
 fn charge_fee(
     state: &mut ValidatedState,
     delta: &mut Delta,
-    fee_info: FeeInfo,
+    fee_info: Vec<FeeInfo>,
     recipient: FeeAccount,
 ) -> Result<(), FeeError> {
-    state.charge_fee(fee_info, recipient)?;
-    delta.fees_delta.extend([fee_info.account, recipient]);
+    for fee_info in fee_info {
+        state.charge_fee(fee_info, recipient)?;
+        delta.fees_delta.extend([fee_info.account, recipient]);
+    }
     Ok(())
 }
 
-/// Validate builder account by verifying signature
+/// Validate builder accounts by verifying signatures. All fees are
+/// verified against signature by index.
 fn validate_builder_fee(proposed_header: &Header) -> Result<(), BuilderValidationError> {
```

Comment:
This should be versioned instead of replaced, pre-marketplace we use old behavior.

Reply:
*I think* this is okay.  To support old behavior, Header v1 logic has been updated to wrap fees and signatures in a `Vec`. So I think this should be ok *in principal*. I'll go over it again to look for implementation errors  though.

Reply:
Oh, I see. I'm seeing `invalid builder fee: Invalid Builder Signature` errors in `demo-native` though so something somewhere doesn't work.

Reply:
We were definitely missing a test for this. I've add a test and it is working as expected. Is it possible that signing is being done incorrectly? 4e3b484546acf27888e9699bdfdeb9a7822b67d3

Reply:
Sorry I wasn't thinking clearly.  c4a9a8edff1ccfff2ef912e4b0d0b53a13ea778a should fix it though.